ctapipe/io/metadata.py

```diff
@@ -0,0 +1,188 @@
+them (as in `Activity.from_provenance()`)
+
+.. code-block:: python
+
+    ref = Reference(
+        contact=Contact(name="Some User", email="user@me.com"),
+        product=Product(format='hdf5', ...),
+        process=Process(...),
+        activity=Activity(...),
+        instrument = Instrument(...)
+    )
+
+    some_astropy_table.meta = ref.to_dict()
+    some_astropy_table.write("output.ecsv")
+
+"""
+import uuid
+from collections import OrderedDict
+
+from astropy.time import Time
+from traitlets import (
+    Enum,
+    Unicode,
+    Int,
+    HasTraits,
+    default,
+    Instance,
+    validate,
+)
+
+__all__ = [
+    "Reference",
+    "Contact",
+    "Process",
+    "Product",
+    "Activity",
+    "Instrument",
+    "write_to_hdf5",
+]
+
+
+class Contact(HasTraits):
+    """ Contact information """
+
+    name = Unicode("unknown")
+    email = Unicode("unknown")
+    organization = Unicode("unknown")
+
+
+class Product(HasTraits):
+    """Data product information"""
+
+    description = Unicode("unknown")
+    creation_time = Unicode()
+    _id = Unicode(help="leave unspecified to automatically generate a UUID")
+    data_category = Enum(["S", "A", "B", "C", "Other"], "Other")
+    data_level = Enum(
+        ["R0", "R1", "DL0", "DL1", "DL2", "DL3", "DL4", "DL5", "DL6", "Other"], "Other"
+    )
+    data_association = Enum(["Subarray", "Telescope", "Target", "Other"], "Other")
+    data_model_name = Unicode("unknown")
+    data_model_version = Unicode("unknown")
+    data_model_url = Unicode("unknown")
+    format = Unicode()
+
+    @default("creation_time")
+    def default_time(self):
+        return Time.now().iso
+
+    @validate("creation_time")
+    def valid_time(selfs, proposal):
+        thetime = Time(proposal["value"])
+        return thetime.iso
+
+    @default("_id")
+    def default_product_id(self):
+        return str(uuid.uuid4())
+
+
+class Process(HasTraits):
+    """ Process (top-level workflow) information """
+
+    _type = Enum(["Observation", "Simulation", "Other"], "Other")
+    subtype = Unicode("")
+    _id = Int()
+
+
+class Activity(HasTraits):
+    """ Activity (tool) information """
+
+    @classmethod
+    def from_provenance(cls, activity: "ActivityProvenance"):
+        return Activity(
+            name=activity["activity_name"],
+            _type="software",
+            _id=activity["activity_uuid"],
+            start=activity["start"]["time_utc"],
+            software_name="ctapipe",
+            software_version=activity["system"]["ctapipe_version"],
+        )
+
+    name = Unicode()
+    _type = Unicode("software")
+    _id = Unicode()
+    start_time = Unicode()
+    software_name = Unicode("unknown")
+    software_version = Unicode("unknown")
+
+    @default("start_time")
+    def default_time(self):
+        return Time.now().iso
+
+
+class Instrument(HasTraits):
+    """ Instrumental Context """
+
+    site = Enum(["CTA-North", "CTA-South", "Other"], "Other")
```

Comment:
Does this mean if I use ctapipe for another observatory, ie magic, I'm forced to use `Other` instead of 'MAGIC'?

Reply:
Yes.  We could add others, though this is "CTA Metadata" the keywords start with the word "CTA" and are defined only for CTA, so it is not intended to be a generic scheme. In this case "site" means which part of CTA, not which observatory site in general (which I guess would be *Roque de los Muchachos* for CTA-N).  However, I do need to add SDMC, since some files will be produced off-site.  Anyhow, this may evolve as the CTA data model evolves, but for now I just want to support what is documented.

Reply:
For files generated for other telescopes, they can define their own other metadata keywords. Probably we should have a top-level one, such as the one defined by OGIP for FITS that tells which instrument you are even talking about( `TELESCOP="CTA"` then imples use the CTA Reference Headers, while `TELESCOP="MAGIC"` would have other metadata)

Reply:
But since we do have some files in ctapipe-extra for other telescopes, and I would like to label them with this scheme, I've added MAGIC, HESS, VERITAS, FACT, and Whipple (even though those are not really "sites"). Note that the real site will come in the Observation Configuration metadata, which is not yet defined (that gives the site name "Paranal", "Roque de los Muchachos", etc, as well as the Earth Location in lon, lat, altitude, etc).  This is just to associate a data product with a part of CTA at the highest-level.